src/ngraph/pass/zero_dim_tensor_elimination.cpp

```diff
@@ -0,0 +1,176 @@
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <memory>
+#include <set>
+
+#include "ngraph/graph_util.hpp"
+#include "ngraph/log.hpp"
+#include "ngraph/op/avg_pool.hpp"
+#include "ngraph/op/broadcast.hpp"
+#include "ngraph/op/constant.hpp"
+#include "ngraph/op/convolution.hpp"
+#include "ngraph/op/max_pool.hpp"
+#include "ngraph/op/pad.hpp"
+#include "ngraph/op/product.hpp"
+#include "ngraph/op/sum.hpp"
+#include "zero_dim_tensor_elimination.hpp"
+
+using namespace ngraph;
+
+#define TI(x) std::type_index(typeid(x))
+
+static std::shared_ptr<Node> build_scalar_constant(std::string val, std::shared_ptr<Node> n)
+{
+    auto cvals = std::vector<std::string>(shape_size(n->get_shape()), val);
+    return std::make_shared<op::Constant>(n->get_element_type(), n->get_shape(), cvals);
+}
+
+static std::shared_ptr<Node> return_zero_const(std::shared_ptr<Node> n)
+{
+    return build_scalar_constant(std::string("0"), n);
+}
+
+static std::shared_ptr<Node> return_one_const(std::shared_ptr<Node> n)
+{
+    return build_scalar_constant(std::string("1"), n);
+}
+
+static std::shared_ptr<Node> return_broadcast(std::shared_ptr<Node> n)
+{
+    AxisSet axes{};
+    for (size_t i = 0; i < n->get_shape().size(); i++)
+    {
+        axes.insert(i);
+    }
+    return std::make_shared<op::Broadcast>(n->get_input_op(1), n->get_shape(), axes);
+}
+
+static std::unordered_map<std::type_index,
+                          std::function<std::shared_ptr<Node>(std::shared_ptr<Node>)>>
+    initialize_const_values_to_ops()
+{
+    return std::unordered_map<std::type_index,
+                              std::function<std::shared_ptr<Node>(std::shared_ptr<Node>)>>({
+        {TI(ngraph::op::AvgPool), return_zero_const},
+        {TI(ngraph::op::MaxPool), return_zero_const},
+        {TI(ngraph::op::Convolution), return_zero_const},
+        {TI(ngraph::op::Sum), return_zero_const},
+        {TI(ngraph::op::Product), return_one_const},
+        {TI(ngraph::op::Pad), return_broadcast},
+    });
+}
+
+static std::unordered_map<std::type_index,
+                          std::function<std::shared_ptr<Node>(std::shared_ptr<Node>)>>
+    ops_to_const_values = initialize_const_values_to_ops();
+
+static std::shared_ptr<Node> get_const_value_for_op(std::shared_ptr<Node> n)
+{
+    const Node& node = *n;
+    auto entry = ops_to_const_values.find(TI(node));
+    return entry == ops_to_const_values.end() ? nullptr : entry->second(n);
+}
```

Comment:
Why not use a virtual `Node::get_const_value_for_op()` whose default returns `nullptr`?